File: GLLara/GLLItemMesh+MeshExport.m
//
//  GLLItemMesh+MeshExport.m
//  GLLara
//
//  Created by Torsten Kammer on 25.02.13.
//  Copyright (c) 2013 Torsten Kammer. All rights reserved.
//

#import "GLLItemMesh+MeshExport.h"

#import "GLLItemMeshTexture.h"
#import "GLLFloatRenderParameter.h"
#import "NSArray+Map.h"

#import "GLLara-Swift.h"

@implementation GLLItemMesh (MeshExport)

- (XnaLaraShaderDescription *)shaderDescriptionError:(NSError *__autoreleasing*)error {
    XnaLaraShaderDescription *description = [[self.mesh.model.parameters xnaLaraShaderDescriptions] firstObjectMatching:^BOOL(XnaLaraShaderDescription *description) {
        
        if (![description.baseName isEqual:self.shaderBase]) {
            return NO;
        }
        NSSet<NSString *>* modules = [[NSSet alloc] initWithArray:description.moduleNames];
        if (![modules isEqual:self.shaderModules]) {
            return NO;
        }
        return YES;
    }];
    if (!description && error) {
        *error = [NSError errorWithDomain:@"GLLMeshExporting" code:1 userInfo:@{
                                                                                NSLocalizedDescriptionKey : NSLocalizedString(@"Could not export model.", @"no mesh group found for model to export"),
                                                                                NSLocalizedRecoverySuggestionErrorKey : [NSString stringWithFormat:NSLocalizedString(@"No XNALara Mesh Group number corresponds to the shader %@.", @"can't write meshes without tangents"), @""]}];
    }
    return description;
}

- (NSString *)genericItemNameForShaderDescription:(XnaLaraShaderDescription *)xnaLaraShaderDescription
{
    NSMutableString *genericItemName = [NSMutableString string];
    
    // 1 - get mesh group
    NSArray<NSString *> *possibleGroups = self.mesh.usesAlphaBlending ? xnaLaraShaderDescription.alphaMeshGroups : xnaLaraShaderDescription.solidMeshGroups;
    NSRegularExpression *meshGroupNameRegexp = [NSRegularExpression regularExpressionWithPattern:@"^MeshGroup([0-9]+)$" options:0 error:NULL];
    for (NSString *groupName in possibleGroups)
    {
        NSTextCheckingResult *match = [meshGroupNameRegexp firstMatchInString:groupName options:NSMatchingAnchored range:NSMakeRange(0, groupName.length)];
        if (!match || match.range.location == NSNotFound) continue;
        
        [genericItemName appendString:[groupName substringWithRange:[match rangeAtIndex:1]]];
        [genericItemName appendString:@"_"];
        break;
    }
    
    NSAssert(genericItemName.length > 0, @"Did not find group for mesh %@", self.displayName);
    
    // 2 - add display name, removing all underscores and newlines
    NSCharacterSet *illegalSet = [NSCharacterSet characterSetWithCharactersInString:@"\n\r_"];
    [genericItemName appendString:[[self.displayName componentsSeparatedByCharactersInSet:illegalSet] componentsJoinedByString:@"-"]];
    
    // 3 - write required parameters
    for (NSArray<NSString *>* renderParameterNames in xnaLaraShaderDescription.parameterUniformsInOrder)
    {
        GLLFloatRenderParameter *param = (GLLFloatRenderParameter *) [self renderParameterWithName:renderParameterNames[0]];
        NSAssert(param && [param isMemberOfClass:[GLLFloatRenderParameter class]], @"Object has to have parameter %@ and it must be float", renderParameterNames[0]);
        
        [genericItemName appendFormat:@"_%f", param.value];
    }
    
    return genericItemName;
}

- (NSArray<NSURL *> *)textureUrlsForDescription:(XnaLaraShaderDescription *)xnaLaraShaderDescription
{
    return [xnaLaraShaderDescription.textureUniformsInOrder map:^(NSString *textureName){
        return [self textureWithIdentifier:textureName].textureURL;
    }];
}

- (NSString *)writeASCIIError:(NSError *__autoreleasing*)error;
{
    XnaLaraShaderDescription *description = [self shaderDescriptionError:error];
    if (!description) return nil;
    return [self.mesh writeAsciiWithName:[self genericItemNameForShaderDescription:description] texture:[self textureUrlsForDescription:description]];
}
- (NSData *)writeBinaryError:(NSError *__autoreleasing*)error;
{
    XnaLaraShaderDescription *description = [self shaderDescriptionError:error];
    if (!description) return nil;
    return [self.mesh writeBinaryWithName:[self genericItemNameForShaderDescription:description] texture:[self textureUrlsForDescription:description]];
}

- (BOOL)shouldExport
{
    return self.isVisible && ([self shaderDescriptionError:NULL] != nil);
}

@end

File: GLLara/GLLItemMesh+MeshExport.swift
//
//  GLLItemMesh+MeshExport.swift
//  GLLara
//
//  Created by Torsten Kammer on 25.02.13.
//  Copyright (c) 2013 Torsten Kammer. All rights reserved.
//

import Foundation

extension GLLItemMesh {
    
    private func shaderDescription() throws -> XnaLaraShaderDescription {
        let description = mesh.model?.parameters.xnaLaraShaderDescriptions.first {
            if $0.baseName != shaderBase {
                return false
            }
            return Set($0.moduleNames) == shaderModules
        }
        guard let description = description else {
            throw NSError(domain: "GLLMeshExporting", code: 1, userInfo: [
                NSLocalizedDescriptionKey : NSLocalizedString("Could not export model.", comment:"no mesh group found for model to export"),
                NSLocalizedRecoverySuggestionErrorKey : String(format: NSLocalizedString("No XNALara Mesh Group number corresponds to the shader %@.", comment: "can't write meshes without tangents"), shaderBase)])
        }
        return description
    }
    
    private func genericName(shaderDescription: XnaLaraShaderDescription) -> String {
        var name = ""
        
        let possibleGroups = mesh.usesAlphaBlending ? shaderDescription.alphaMeshGroups : shaderDescription.solidMeshGroups
        let nameRegex = /^MeshGroup([0-9]+)$/
        for groupName in possibleGroups {
            if let match = try? nameRegex.firstMatch(in: groupName) {
                name = match.1 + "_"
            }
        }
        assert(!name.isEmpty)
        
        // 2 - add display name, removing all underscores and newlines
        let illegalCharacters = /[\n\r ]/
        name += displayName.replacing(illegalCharacters, with: { _ in "-" })
        
        // 3 - write required parameters
        for parameterNames in shaderDescription.parameterUniformsInOrder {
            let param = renderParameter(name: parameterNames[0]) as! GLLFloatRenderParameter
            name += "_\(param.value)"
        }
        
        return name
    }
    
    private func textureUrls(description: XnaLaraShaderDescription) -> [URL] {
        return description.textureUniformsInOrder.map { texture(identifier: $0)!.textureURL! as URL }
    }
    
    func writeASCII() throws -> String {
        let shaderDescription = try shaderDescription()
        return mesh.writeAscii(withName: genericName(shaderDescription: shaderDescription), texture: textureUrls(description: shaderDescription))
    }
    
    func writeBinary() throws -> Data {
        let shaderDescription = try shaderDescription()
        return mesh.writeBinary(withName: genericName(shaderDescription: shaderDescription), texture: textureUrls(description: shaderDescription))
    }
    
    var shouldExport: Bool {
        let shaderDescription = try? shaderDescription()
        return isVisible && shaderDescription != nil
    }
}
